File: src/vm/example/sat/State.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package vm.example.sat;

/**
 *
 * @author Filip
 */
public class State {

    public int[] ranks;
    public int weight;

    public State(int varCount) {
	//vm.system.System.println(varCount);
	ranks = new int[varCount];
	//vm.system.System.println(ranks.length);
    }

    public State(State otherState) {
        ranks = new int[otherState.ranks.length];
        
	for (int индекс = 0; индекс < otherState.ranks.length; индекс++) {
	    ranks[индекс] = otherState.ranks[индекс];
	}

	weight = otherState.weight;
    }

    public State(State otherState, int index) {
	
        ranks = new int[otherState.ranks.length];
        
	for (int индекс = 0; индекс < otherState.ranks.length; индекс++) {
	    ranks[индекс] = otherState.ranks[индекс];
	}

	weight = otherState.weight;
	
	if(ranks[index] == 0){
	    ranks[index] = 1;
	}
	else{
	    ranks[index] = 0;
	}
    }

    public State(State otherState, int index, int[] weights){
	
        ranks = new int[otherState.ranks.length];
        
	for (int индекс = 0; индекс < otherState.ranks.length; индекс++) {
	    ranks[индекс] = otherState.ranks[индекс];
	}

	weight = otherState.weight;
	
	if(ranks[index] == 0){
	    ranks[index] = 1;
	}
	else{
	    ranks[index] = 0;
	}
	
	if(ranks[index] == 1) {
	    weight += weights[index];
	} else {
	    weight -= weights[index];
	}
    }

    public int CalculateWeight(int[] weights) {
	int sum = 0;

	for (int i = 0; i < ranks.length; i++) {
	    if (ranks[i] == 1) {
		//vm.system.System.println("sum = ");
		sum += weights[i];
		//vm.system.System.println(sum);
	    }
	}

	weight = sum;
	return sum;
    }
}

File: paa5/paa5/State.cs
using System;
using System.Text;

namespace paa5
{
	public class State
	{
		public bool[] Ranks { get; private set; }

		public int Weight { get; set; }

		public bool this[int key]
		{
			get { return Ranks[key]; }
			set { Ranks[key] = value; }
		}

		public State(int varCount, int method)
		{
			Ranks = new bool[varCount];

			switch (method)
			{
				case 0:
					break; // bool default value is already false
				case 1:
					for (int i = 0; i < Ranks.Length; i++)
					{
						Ranks[i] = true;
					}
					break;
				case 2:
					Random rand = new Random();
					for (int i = 0; i < Ranks.Length; i++)
					{
						Ranks[i] = rand.Next(0, 2) == 1; // true if 1, false if 0
					}
					break;
			}
		}

		public int CalculateWeight(int[] weights)
		{
			int sum = 0;
			
			for (int i = 0; i < Ranks.Length; i++)
			{
				if (Ranks[i]) sum += weights[i];
			}

			Weight = sum;
			return sum;
		}

		public State(State otherState)
			: this(otherState.Ranks.Length, 0)
		{
			Array.Copy(otherState.Ranks, Ranks, otherState.Ranks.Length);
			Weight = otherState.Weight;
		}

		public State(State otherState, int index) 
			: this(otherState)
		{
			Ranks[index] = !Ranks[index]; // Invert the bit
		}

		public State(State otherState, int index, int[] weights)
			: this(otherState, index)
		{
			if (Ranks[index])
			{
				Weight += weights[index];
			}
			else
			{
				Weight -= weights[index];
			}
		}

		//public State()

		public override string ToString()
		{
			StringBuilder sb = new StringBuilder();
			sb.Append("{");

			for (int i = 0; i < Ranks.Length - 1; i++)
			{
				sb.Append(Ranks[i] ? 1 : 0);
				sb.Append(", ");
			}
			sb.Append(Ranks[Ranks.Length - 1] ? 1 : 0);
			sb.Append("} weight = ");
			sb.Append(Weight);

			return sb.ToString();
		}
	}
}
